client/src/GameState.ts

```diff
@@ -126,6 +205,157 @@ export class GameState {
+        task.range
+            .slice()
+            .reverse()
+            .forEach((lineIndex) => {
+                if (task.axis === LineCheckAxis.Row) {
+                    const offset =
+                        task.direction * (linesToFall.get(lineIndex) || 0);
+                    if (offset === 0) {
+                        return;
+                    }
+                    const newRow = lineIndex + offset;
+
+                    // swap line with the offset destination line
+                    // the old line should always be null since it's cleared/propagated
+                    for (let col = 15; col < 25; col++) {
+                        this.board[newRow][col] = this.board[lineIndex][col];
+                        this.board[lineIndex][col] = null;
+                        const newMonomino = this.board[newRow][col];
+                        if (newMonomino) {
+                            newMonomino.position[0] = newRow;
+                        }
+                    }
+                } else if (task.axis === LineCheckAxis.Column) {
+                    const offset =
+                        task.direction * (linesToFall.get(lineIndex) || 0);
+                    if (offset === 0) {
+                        return;
+                    }
+                    const newCol = lineIndex + offset;
+
+                    for (let row = 15; row < 25; row++) {
+                        this.board[row][newCol] = this.board[row][lineIndex];
+                        this.board[row][lineIndex] = null;
+                        const newMonomino = this.board[row][newCol];
+                        if (newMonomino) {
+                            newMonomino.position[1] = newCol;
+                        }
+                    }
+                }
+            });
+    }
+
+    private prepareLinesToFall(
+        task: LineCheckTask,
+        linesToClear: Array<number>
+    ): Map<number, number> {
+        const linesOffset = new Map<number, number>();
+        let offset = 0;
+        if (linesToClear.length > 0) {
+            task.range
+                .slice()
+                .reverse()
+                .forEach((lineIndex) => {
+                    if (lineIndex === linesToClear?.at(0)) {
+                        // this line is an empty, cleared line. all lines above should fall with one more offset
+                        offset += 1;
+                        linesToClear.push(linesToClear.shift()!);
+                    } else {
+                        linesOffset.set(lineIndex, offset);
+                    }
+                });
+        }
+        return linesOffset;
+    }
+
+    private scanLinesToClear(task: LineCheckTask): Array<number> {
+        const linesToClear = [];
+        if (task.axis === LineCheckAxis.Row) {
+            for (const row of task.range) {
+                // scan through each row
+                let canClear = true;
+                for (let col = 15; col < 25; col++) {
```

Comment:
These should be expressed as `WALL_SIZE` and `BOARD_SIZE - WALL_SIZE` instead of being magic numbers